File: 77.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        if k == 1:
            return [[i] for i in range(1,n+1)]
        elif k == n:
            return [[i for i in range(1,n+1)]]
        else:
            res = []
            res += self.combine(n-1, k)
            part = self.combine(n-1,k-1)
            for item in part:
                item.append(n)
            res += part
            return res

File: 77.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> nums;
        for (int i = 1; i <= n; i++){
            nums.push_back(i);
        }
        vector<vector<int>> res;
        vector<int> out;
        dfs(nums, res, out, k, 0);
        return res;
    }
    void dfs(vector<int>& nums, vector<vector<int>>& res, vector<int>& out, int k, int index){
        if (out.size() == k){
            res.push_back(out);
            return;
        }
        for (int i = index; i < nums.size(); i++){
            out.push_back(nums[i]);
            dfs(nums, res, out, k, i+1);
            out.pop_back();
        }
        return ;
    }
};
